lib/spack/spack/package.py

```diff
@@ -1565,19 +1566,20 @@ def run_test(self, exe, options=[], expected=[], status=None,
 
         try:
             output = runner(*options, output=str.split, error=str.split)
-            assert not status, 'Expected execution to fail'
-
-            for check in expected:
-                cmd = ' '.join([exe] + options)
-                msg = "Expected '%s' in output of `%s`" % (check, cmd)
-                msg += '\n\nOutput: %s' % output
-                assert check in output, msg
 
+            can_pass = None in status or 0 in status
+            assert can_pass, 'Expected execution to fail'
         except ProcessError as err:
             output = str(err)
-            status_msg = 'exited with status {0}'.format(status)
-            if status_msg not in output:
-                raise  # re-raise the same error/traceback
+            code = int(re.findall(r'[0-9]+', output.split('\n')[0])[0])
```

Comment:
Can we make this more robust? I'm worried that this version will pick up any numerical instance in the output, whether or not it's an error message.

Maybe we could do

```
codes = '|'.join([s for s in status if s])
re.search(r'exited with status (%s)' % codes', output)
```